### DISCUSSION TITLE
How can pressure driven demand and leaks be represented?

### DISCUSSION THREAD

**Discussion by GeertD** [e2nIEE/pandapipes#410] [Q&A]
Hi,

I'm evaluating Pandapipes vs Epanet (WNTR python interface) for analyzing water/gas networks.
Pandapipes looks to have a more friendly api and is more general (also for gas, compressible fluids, ...).

A common use case is **pressure dependant demand at the nodes**.
Typically: flowrate = C * sqrt(pressure) ; with C the discharge coefficient.
Is called an emitter in Epanet, used to model e.g. leaks, ... .

In Pandapipes, I can't find a straightforward option for this.
I assume this can be modelled with a node + valve + common sink for all these emitters, but that looks cumbersome.

Any option to implement this at the node level?

Best regards,

Geert

**Comment by jkisse** [comment]
Hi Geert,

we will talk about this in the developer team, but I have a first idea. 

Generally speaking, what you want to do is modifying a component parameter (mdot_kg_per_s of the sinks) in dependence of an output variable of the pipe flow calculation (nodal pressure), which leads to a circularity.

We have a similar behaviour in power grids, e. g. when the tap position of transformers ("RONT") could be set in dependence of the voltage. For this, pandapower provides the concept of [controllers](https://pandapower.readthedocs.io/en/latest/control.html). Controllers can manipulate components' parameters during the power / pipe flow calculation. A basic controller from pandapower can be found [here](https://github.com/e2nIEE/pandapower/blob/develop/pandapower/control/controller/characteristic_control.py). 

In pandapipes, the run_control procedure is already implemented, but currently only so called "multi net" controllers are available (for a different purpose). Nonetheless, I think pandapower's "characteristic controller" could be quite easily adapted to fit your needs.

I hope this first idea helps you in developing your approach.

Best regards,
Jolando

**Comment by GeertD** [comment]
Hi Jolando,

Thank you for the quick reply.
I'll have a look into these controllers.
Although these might create a lot of overhead for a network with many
similar leaks (like pipes with orifices, irrigation, ...).

The way Epanet seems to tackle this is by reversing the relation.
The user specifies outflow as a function of pressure at the node: Q = f(p),
typically Q = C * sqrt(p).
Epanet reverses this: p = (Q/C)^2, and they add the node pressure drop to
the pipe pressure drop. They give Q an initial guess and calculate the
network until balanced.
Details in Epanet manual, under hydraulics.


Best regards,

Geert

**Comment by jkisse** [comment]
Hi Geert,
actually, the overhead is quite small, since one controller can be used for all sinks with the same characteristic function. I have tried it with the pandapower "CharacteristicController" and it seems to work well. Here is a little example:

```
import pandapipes as ps
from pandapipes import networks as nw
from pandapower.control.controller.characteristic_control import CharacteristicControl
from pandapipes.control import run_control

# load an example network and prepare it:
net = nw.water_combined_mixed()
net.sink.mdot_kg_per_s = 0
net.source.in_service = False

# the characteristic function (pressure dependant demand)
def mdot_from_p(p):
    C = 0.1  # scaling factor
    return C*(p**0.5)

output_element = 'sink'
output_variable = 'mdot_kg_per_s'
output_element_index = net.sink.index
input_element = 'res_junction'
input_variable ='p_bar'
input_element_index = net.sink.junction.values
characteristic = mdot_from_p

CharacteristicControl(net, output_element, output_variable, output_element_index, input_element,
                 input_variable, input_element_index, characteristic)

run_control(net)  # --> pipeflow calculation that applies the CharacteristicControl during the simulation

#results:
print(net.res_junction.p_bar ** 0.5)
print(net.res_sink.mdot_kg_per_s /0.1)  # 0.1 comes from the C factor
```
The calculation of the example takes ~30 ms on my system.

It would be very interesting for us to hear if this is helpful for your question.

Best regards,
Jolando
